### ISSUE TITLE
Any plans for Linux Build?

### ISSUE THREAD

**Issue by JahlomP** [trueinteractions/tint2#98]
Any plans for Linux Build? Some of us will like to contribute and even build upon it but we are Linux guys.

**Comment by trevorlinton** [comment]
Hi @JahlomP 

We're currently working on iOS/Android ports at the moment so linux support isn't on the radar.  Part of this isn't to snub linux platforms but its fairly confusing as to how to approach the issue.

1. Linux does not (generally speaking) have one architecture to support.  Linux could hypothetically run on arm or x86 based platforms.
2. Linux isn't a GUI toolkit; and there are a very large selection of toolkits to choose from, Gnome and QT come to mind as the first to support; however both do not take care of "UI Boundary" problems.
3. Say if we want to store user preferences or abide by not having an application appear in the dock, but as a background application.  Application registration and lifecycles do not have a standard, or even a known approach on major toolkits like QT/Gnome.  This significantly reduces the effectiveness of using something such as Tint simply because it become unweidly to cover so many corner cases and different "Dock" or "Taskbar" systems used by such a wide variety of programmers.
4.  Non-standard window managers.  To further complicate issues window managers can be swapped out independently of Gnome and QT (and in quite a few cases may completely ignore the Gnome or QT contracts).  This in our initial testing had lead to a larger than we can tackle set of bugs as to how to handle simple things such as minimization, maximization and application top-foremost issues. 
5.  No build in gatekeeper or sandboxing.  Sadly linux does not have a standard gatekeeping or sandbox mechanism for desktop applications.  Meaning, there's no way of establishing boundaries of where we can and can't write to beyond standard posix users, groups and permission sets. 
6. No standardized application packaging mechanism or manifest format.  While plenty of these do exist, there's no standard, and no way of ensuring how we package an application will be respected by the desktop environment receiving it. The only alternative we could find was supporting up to 12 different packaging formats all with their own inherit issues.
7. Composition is also much of a quandary, the way OpenGL is implemented on Linux desktop environments is HIGHLY volatile.  Some implementations require using toolkits to receive OpenGL views, while some require directly using IPC to at the composite manager to receive OpenGL views.  And then some have specific API's at the window management (and in some cases, the window server) to get open gl views.   The lifecycle of these vary as well and have significant challenges to adopt.
8. There is no sunset clause for various packages and API's.  While modularization and customization may be great for hacking around on a linux platform it makes it very difficult for a person in our position to determine what level of support at what API version we could support for each individual module used when creating a desktop environment.  There's no definitive way to say which toolset we should be supporting and to what version should we retire support.   This creates a maintenance headache as you might imagine.
9.  We can't find any documentation or best practices on the subject as the desktop environment for linux (has been, and continues to be) a volatile space that seems fairly controversial and opinionated.  This makes it difficult from a political stand point.  In effect we need to "pick our side" and go forward with an approach that we hope will be successful, completely unrelated from the technical merits of the toolchain, toolkit, libraries and performance.
10. There is no standard reflective API that exists. Much of the work Apple, Microsoft and Sun Microsystems have done is to create API's that allow for UI widgets to cross between various application boundaries to support things like drag-and-drop or OLE/COM/NSObject/OSX OBJC Runtime/.NET. While ports for nearly ALL of these technologies exist on Linux, none of them have been realized to their full effect (e.g., while we could use java or the mono runtime, there's no real usefulness if QT doesn't support their base UI component, such as in WPF, and even if we did, support for UI related activities in these reflective languages is minimal at best).

So there's  ALOT of caveats with development on a Linux Desktop environment purely because there isn't a clear leader in this space.  The best we can hope for is to pick a complete platform (perhaps,  Ubuntu?..) and ONLY support ubuntu on a specific architecture (such as X86), with the additional caveat that if you've customized the environment beyond what Ubuntu allows for in its normal settings, the behavior of your application becomes undefined.

It appears even Ubuntu is aware of this and has been pushing devs on Linux to just create an HTML application, which it seems a lot of users are going that route.

Thoughts?

**Comment by JahlomP** [comment]
Wow. That is a lot. I will forward this to my patners and get back to you
if we have something. Thanks

**Comment by trevorlinton** [comment]
FYI, for those interested in contributing or otherwise, we've added support for linux and GTK, its now to a place where the bridge is functional and the event loop is pretty solid (and it builds!), see:
https://github.com/trueinteractions/tint2/wiki/Linux-Support-Status

For more information.

**Comment by Webkitnote** [comment]
Sorry for necro-bumping. Maybe it's stupid to ask, but the link you provided, suggests that there are actually people voluntarily attempting to make Tint work on Linux. So there will be be support for Linux, sooner or later, right?

**Comment by trevorlinton** [comment]
@Webkitnote Yes, we're looking at supporting linux soon.  At the moment were trying to ensure our compiler and win/osx versions are as stable as possible, then the next support will come to Ubuntu and Android platforms.

**Comment by Webkitnote** [comment]
Would there be a wild guess from your side when such thing could happen? 3 months? 6 months? A year?
